app/models/container_deployment.rb

```diff
@@ -6,5 +6,355 @@ class ContainerDeployment < ApplicationRecord
+        "githubOrganizations" => authentication.github_organizations
+      }
+    when "AuthenticationGoogle"
+      options = {
+        "clientID"     => authentication.userid,
+        "clientSecret" => authentication.password,
+        "hostedDomain" => authentication.google_hosted_domain,
+        "challenge"    => "false"
+      }
+    when "AuthenticationHtpasswd"
+      options = {
+        "filename" => "/etc/origin/master/htpasswd"
+      }
+    when "AuthenticationLdap"
+      options = {
+        "attributes"   => {"id"                => authentication.ldap_id,
+                           "email"             => authentication.ldap_email,
+                           "name"              => authentication.ldap_name,
+                           "preferredUsername" => authentication.ldap_preferred_user_name},
+        "bindDN"       => authentication.ldap_bind_dn,
+        "bindPassword" => authentication.password,
+        "ca"           => authentication.certificate_authority,
+        "insecure"     => authentication.ldap_insecure.to_s,
+        "url"          => authentication.ldap_url
+      }
+    when "AuthenticationOpenId"
+      options = {
+        "clientID"                       => authentication.userid,
+        "clientSecret"                   => authentication.password,
+        "claims"                         => {"id" => authentication.open_id_sub_claim},
+        "urls"                           => {"authorize" => authentication.open_id_authorization_endpoint, "toekn" => authentication.open_id_token_endpoint},
+        "challenge"                      => "false",
+        "openIdExtraAuthorizeParameters" => authentication.open_id_extra_authorize_parameters,
+        "openIdExtraScopes"              => authentication.open_id_extra_scopes
+      }
+    when "AuthenticationRequestHeader"
+      options = {
+        "challengeURL"                          => authentication.request_header_challenge_url,
+        "loginURL"                              => authentication.request_header_login_url,
+        "clientCA"                              => authentication.certificate_authority,
+        "headers"                               => authentication.request_header_headers,
+        "requestHeaderPreferredUsernameHeaders" => authentication.request_header_preferred_username_headers,
+        "requestHeaderNameHeaders"              => authentication.request_header_name_headers,
+        "requestHeaderEmailHeaders"             => authentication.request_header_email_headers
+      }
+    end
+    {'name' => "example_name", 'login' => "true", 'challenge' => "true", 'kind' => authentication.authtype}.merge! options
+  end
+
+  def extract_public_ip_or_hostname(deployment_node)
+    return deployment_node.address if deployment_node.vm_id.nil?
+    vm = deployment_node.vm
+    hostname = vm.hostnames
+    if hostname.empty?
+      ips = vm.hardware.ipaddresses
+      ips.each do |ip|
+        return ip if public_ip? ip
```

Comment:
We generally don't use [seattle.rb style, leaving off parentheses unless required](https://heartmindcode.com/2013/09/05/parentheses-in-ruby/).  I have nothing against it here but we don't use it elsewhere.